Fixed #373 -- Added CompositePrimaryKey. (#18056)
django/django: docs/howto/composite-primary-key.txt

In most cases, a single-field primary key should suffice.
In database design, however, defining a primary key that consists of multiple
fields is sometimes necessary.

To define a composite primary key, add the ``models.CompositePrimaryKey``
field to your model:

.. code-block:: pycon

    class Tenant(models.Model):
        pass

    class User(models.Model):
        pk = models.CompositePrimaryKey("tenant_id", "id")
        tenant = models.ForeignKey(Tenant, on_delete=models.CASCADE)
        id = models.IntegerField()

This will instruct Django to create a composite ``PRIMARY KEY`` constraint
instead of a single-field primary key.

It's possible to define a composite primary key consisting of a single field.
This is practically the same as setting ``primary_key=True`` on the field.

.. code-block:: pycon

    class Model(models.Model):
        pk = models.CompositePrimaryKey("id")
        id = models.IntegerField()

Migrating to a composite primary key
====================================

As of now, Django doesn't offer built-in support for the migration of primary
keys.

This means once a table is created, its primary key cannot be changed via

> Django's migration operations.

Author: Sure, we can document it!

Unfortunately, changing a table's primary key is a complex and hairy procedure.
Django doesn't really support changing the pk via migrations, composite pk is no exception.

> Django's migration operations.

Reviewer: One thing we might also be able to do here is provide a more readable error when running `makemigrations`. Do we have anything like this for normal pk changes?

> Django's migration operations.

Author: There's no special error, it usually just fails with a db error.

> Django's migration operations.

Reviewer 2: So this case didn't fail when applying the migration and perhaps it might be nice to have alterfield on CompositePrimaryKeys to fail loudly

> maybe we should (or already do) have a ticket for this in the normal pk case?

Agree a ticket would be great, we should be specific on what this is. 
We do have some tickets that are in this [rough "area"](https://code.djangoproject.com/query?component=Migrations&description=~primary+key&status=assigned&status=new&order=id&desc=1&col=id&col=summary&col=status&col=component&col=owner&col=type&col=version).

> Django's migration operations.

Author: Afaik, it's the same "no operation" behaviour as running an AlterField with `primary_key=False` or `primary_key=True`.
So if we want an error, it should also take that into consideration .
I'm all for making this as user friendly as possible, I could add this but maybe in another PR. I can create a ticket for it tomorrow.

> Django's migration operations.

Author: I adjusted the docs.

> Django's migration operations.

Author: https://code.djangoproject.com/ticket/35589#ticket

If you would like to migrate an existing table from a single primary key to a
composite primary key, follow your database backend's instructions to do so.

Once the composite primary key is in place, add the ``CompositePrimaryKey``
field to your model. This allows Django to recognize and handle the composite
primary key appropriately.

.. note::

    While migration operations (e.g. ``AddField``, ``AlterField``) on
    primary key fields are not supported, ``makemigrations`` will still detect
    changes.

    In order to avoid errors, it's recommended to apply such migrations with
    ``--fake``.

How to define a composite foreign key
=====================================

> At the moment, ``ForeignKey`` doesn't support composite primary keys.

Reviewer: This would be all relationship fields.

> At the moment, ``ForeignKey`` doesn't support composite primary keys.

Author: I adjusted the docs.

For the time being, the ``ForeignObject`` field can be used as an alternative:

.. code-block:: pycon

    class Comment(models.Model):
        pk = models.CompositePrimaryKey("tenant_id", "id")
        tenant = models.ForeignKey(Tenant, on_delete=models.CASCADE)
        id = models.IntegerField()
        user_id = models.IntegerField()
        user = models.ForeignObject(
            User,
            on_delete=models.CASCADE,
            from_fields=("tenant_id", "user_id"),
            to_fields=("tenant_id", "id"),

>         )

Reviewer 2: > We were thinking of using ForeignObject temporarily, until support for proper composite foreign keys is added. Would that be okay?

Yes that's fine I think 👍

>         )

Author: I added some docs, let me know if that's ok.

This is practically the same as setting ``primary_key=True`` on the field.

.. code-block:: pycon

    class Model(models.Model):
        pk = models.CompositePrimaryKey("id")
        id = models.IntegerField()

Migrating to a composite primary key
====================================

As of now, Django doesn't offer built-in support for the migration of primary
keys.

This means once a table is created, its primary key cannot be changed via
Django's migration operations, regardless of whether the table was created
with a regular primary key or a composite primary key.

If you would like to migrate an existing table from a single primary key to a
composite primary key, follow your database backend's instructions to do so.

Once the composite primary key is in place, add the ``CompositePrimaryKey``
field to your model. This allows Django to recognize and handle the composite
primary key appropriately.

.. note::

    While migration operations (e.g. ``AddField``, ``AlterField``) on
    primary key fields are not supported, ``makemigrations`` will still detect
    changes.

    In order to avoid errors, it's recommended to apply such migrations with

>     ``--fake``.

Reviewer: Maybe we should recommend [`SeparateDatabaseAndState`](https://docs.djangoproject.com/en/5.0/ref/migration-operations/#separatedatabaseandstate) here.

Author: Yes, I added a sentence.

=====================================
How to define a composite primary key
=====================================

.. versionadded:: 5.2

In Django, each model has a primary key. By default, this primary key consists
of a single field.

In most cases, a single primary key should suffice. In database design,
however, defining a primary key consisting of multiple fields is sometimes
necessary.

To define a composite primary key, set the model's ``pk`` to an instance of
``CompositePrimaryKey``:

> .. code-block:: pycon

Author: Ah yes, I mixed it up, thanks for catching this!

        reference = models.CharField(max_length=20, primary_key=True)

    class OrderLineItem(models.Model):
        pk = models.CompositePrimaryKey("product_id", "order_id")
        product = models.ForeignKey(Product, on_delete=models.CASCADE)
        order = models.ForeignKey(Order, on_delete=models.CASCADE)
        quantity = models.IntegerField()

This will instruct Django to create a composite primary key
(``PRIMARY KEY (product_id, order_id)``) when creating the table.

Composite primary keys are represented with ``tuple``\s:

.. code-block:: pycon

    >>> product = Product.objects.create(name="apple")
    >>> order = Order.objects.create(reference="A755H")
    >>> item = OrderLineItem.objects.create(product=product, order=order, quantity=1)
    >>> item.pk
    (1, "A755H")

You can assign a ``tuple`` to a composite primary key. This sets the associated
field values.

.. code-block:: pycon

    >>> item = OrderLineItem(pk=(2, "B142C"))
    >>> item.pk
    (2, "B142C")
    >>> item.product_id
    2
    >>> item.order_id
    "B142C"

Or filter composite ``pk``\s by ``tuple``\s:

.. code-block:: pycon

    >>> OrderLineItem.objects.filter(pk=(1, "A755H")).count()
    1

> 

Reviewer 3: I like it!

a single operation.

Composite primary keys and relations
====================================

:ref:`Relationship fields <relationship-fields>`, including
:ref:`generic relations <generic-relations>` do not support composite primary
keys.

For example, given the ``OrderLineItem`` model, we cannot do:

.. code-block:: pycon

    class Foo(models.Model):
        item = models.ForeignKey(OrderLineItem, on_delete=models.CASCADE)

Because ``ForeignKey`` currently cannot reference models with composite primary
keys.

To work around this limitation, ``ForeignObject`` can be used as an
alternative:

.. code-block:: pycon

    class Foo(models.Model):
        item_order_id = models.IntegerField()
        item_product_id = models.CharField(max_length=20)
        item = models.ForeignObject(
            OrderLineItem,
            on_delete=models.CASCADE,
            from_fields=("item_order_id", "item_product_id"),
            to_fields=("order_id", "product_id"),
        )

``ForeignObject`` is much like ``ForeignKey``, except that it doesn't create
any columns (e.g. ``item_id``), foreign key constraints or indexes in the
database.

.. warning::

>     ``ForeignObject`` is an internal API. Use it at your own risk.

Author: @sarahboyce says we shouldn't set expectations to the reader: https://github.com/django/django/pull/18056#discussion_r1724948144

>     ``ForeignObject`` is an internal API. Use it at your own risk.

Reviewer 2: For this one, I think the PR/feature is big enough it deserves it's own ticket.
I think an accepted ticket would mean we've committed to support it, then I think we can say that we plan to add this in, but I would avoid words like "soon"

Also maybe
```
    ``ForeignObject`` is an internal API. This means it is not covered by our
    :ref:`deprecation policy <internal-release-deprecation-policy>`.
```

>     ``ForeignObject`` is an internal API. Use it at your own risk.

Author: I agree, we should limit the scope of [#373](https://code.djangoproject.com/ticket/373) (see my [comment](https://code.djangoproject.com/ticket/373#comment:197)).

Composite primary keys and database functions
=============================================

Many database functions only accept a single expression.

.. code-block:: sql

    MAX("order_id")  -- OK
    MAX("product_id", "order_id")  -- ERROR

In consequence, they cannot be used with composite primary key references
as they are composed of multiple column expressions.

> .. code-block:: pycon
> 
>     Max("order_id")  -- OK
>     Max("pk")  -- ERROR

Reviewer 2: ```suggestion
.. code-block:: python

    Max("order_id")  # OK
    Max("pk")  # ERROR
```

Composite primary keys and admin
================================

> The admin doesn't support models with composite primary keys at this time.

Author: According to @sarahboyce , we shouldn't: https://github.com/django/django/pull/18056#discussion_r1724948144

> The admin doesn't support models with composite primary keys at this time.

Reviewer 2: As a expect (:crossed_fingers:) we will merge in the admin support 7c59f8e69cd9c47a77aa92158fc247ab5f9b6507 very soon after, maybe we don't need to document that it's not supported in this pr.
We could add it to this PR as a separate commit (perhaps once reviews of this has settled)

> The admin doesn't support models with composite primary keys at this time.

Author: We gotta merge generic FK support first after this because the admin uses generic FKs. I have a proof of concept ready but we need to iron it out first. What I'm trying to say is it won't be very soon after, but maybe 1-2 months if everything goes according to plan.

> The admin doesn't support models with composite primary keys at this time.

Reviewer 2: I have made a [suggestion here](https://github.com/django/django/pull/18056#discussion_r1839644304)

    >>> item = OrderLineItem(pk=(2, "B142C"))
    >>> item.pk
    (2, "B142C")
    >>> item.product_id
    2
    >>> item.order_id
    "B142C"

Or filter composite ``pk``\s by ``tuple``\s:

.. code-block:: pycon

    >>> OrderLineItem.objects.filter(pk=(1, "A755H")).count()
    1

Migrating to a composite primary key
====================================

Django doesn't support migrating to, or from, a composite primary key after the
table is created.

If you would like to migrate an existing table from a single primary key to a
composite primary key, follow your database backend's instructions to do so.

Once the composite primary key is in place, add the ``CompositePrimaryKey``
field to your model. This allows Django to recognize and handle the composite
primary key appropriately.

While migration operations (e.g. ``AddField``, ``AlterField``) on
primary key fields are not supported, ``makemigrations`` will still detect
changes.

In order to avoid errors, it's recommended to apply such migrations with
``--fake``.

Alternatively, :class:`.SeparateDatabaseAndState` may be used to execute
the backend-specific migrations and Django-generated migrations in
a single operation.

> How to define a composite foreign key
> =====================================
> 
> At the moment, relationship fields (``ForeignKey``, ``OneToOneField``,
> ``ManyToManyField``) don't support composite primary keys.
> 
> So, given the ``OrderLineItem`` model, we cannot do:

Reviewer 2: ```suggestion
Composite primary keys and relations
====================================

:ref:`Relationship fields <relationship-fields>`, including
:ref:`generic relations <generic-relations>` do not support composite primary
keys.

For example, given the ``OrderLineItem`` model, we cannot do:
```
(for this suggestion to work, I think we need to add a `.. _relationship-fields` link)

I think let's include the mention of generic relations here and remove the `Composite primary keys and generic relations` (as I don't think it deserves it's own separate section)

.. code-block:: pycon

    class Foo(models.Model):
        item = models.ForeignKey(OrderLineItem, on_delete=models.CASCADE)

> Because ``ForeignKey`` cannot reference models with composite primary keys
> *yet*.

Reviewer 2: ```suggestion
Because ``ForeignKey`` cannot reference models with composite primary keys.
```
I think we avoid implying work is coming round the corner in docs, in case it sets expectations to the reader.
We can say "currently cannot reference" if you wish

To work around this limitation, ``ForeignObject`` can be used as an
alternative:

.. code-block:: pycon

    class Foo(models.Model):
        item_order_id = models.IntegerField()
        item_product_id = models.CharField(max_length=20)
        item = models.ForeignObject(
            OrderLineItem,
            on_delete=models.CASCADE,
            from_fields=("item_order_id", "item_product_id"),
            to_fields=("order_id", "product_id"),
        )

> ``ForeignObject`` is much like ``ForeignKey``,
> except that it doesn't create any columns (e.g. ``item_id``), foreign key
> constraints or indexes in the database.

Reviewer 2: ```suggestion
``ForeignObject`` is much like ``ForeignKey``, except that it doesn't create
any columns (e.g. ``item_id``), foreign key constraints or indexes in the
database.
```
(nit on line length)

.. warning::

    ``ForeignObject`` is an internal API. The work to provide a public API for
    composite foreign keys is already under way. Use it at your own risk.

Composite primary keys and generic relations
============================================

The composite primary key feature is still in its infancy, it doesn't
support generic relations yet.

The work to add composite primary key support to generic relations
is already under way.

Composite primary keys and database functions
=============================================

> The database functions may not work with composite primary keys.
> 
> This is because most database functions accept a single field only.

Reviewer 2: ```suggestion
Many database functions only accept a single field.
```

.. code-block:: sql

    COUNT("order_id")  -- OK
    COUNT("product_id", "order_id")  -- ERROR

> So, unless the database function accepts multiple fields, don't
> use it with a composite primary key directly.

Reviewer 2: ```suggestion
These database functions cannot be used with composite primary keys, as
composite primary keys represent multiple fields.
```
Not sure "represent" is the best word :thinking: might be able to say "are a proxy to"

> =====================================

Reviewer 4: To me, this document seems like it should be in 'topics' rather than 'how-to'.

Reviewer 2: Agreed, @csirmazbendeguz do you mind making this update?

Author: I moved it 👍

            to_fields=("order_id", "product_id"),
        )

``ForeignObject`` is much like ``ForeignKey``, except that it doesn't create
any columns (e.g. ``item_id``), foreign key constraints or indexes in the
database.

.. warning::

    ``ForeignObject`` is an internal API. This means it is not covered by our
    :ref:`deprecation policy <internal-release-deprecation-policy>`.

Composite primary keys and database functions
=============================================

Many database functions only accept a single expression.

.. code-block:: sql

    MAX("order_id")  -- OK
    MAX("product_id", "order_id")  -- ERROR

As a consequence, they cannot be used with composite primary key references as
they are composed of multiple column expressions.

.. code-block:: python

    Max("order_id")  # OK
    Max("pk")  # ERROR

Composite primary keys in forms
===============================

:meth:`~django.db.models.Model.save()` doesn't support updating primary
keys (composite or not).

This means using :class:`~django.forms.ModelForm` with a composite
primary key may lead to unexpected behavior.

To avoid issues, make sure :attr:`~django.db.models.Field.editable`

> is set to ``False`` on all primary key fields.

Reviewer: We can probably raise an error if this isn't the case - maybe using the checks framework.

Author: Yeah we could add a check to force composite PK fields to be `editable=False`.

I'm not sure if we should. This is an issue with regular primary keys as well: https://code.djangoproject.com/ticket/2259 . So if we force this for composite PKs , shouldn't we force it for regular PKs too?

Reviewer: I had a skim through that issue and it seems the sticking point is creation vs updates - it should be possible to set the pk values during create and not after that. I'm not sure if Django has anything to handle that cleanly. If we don't, I'm happy to say this is out of scope (and in scope of 2259 instead).

> :meth:`~django.db.models.Model.save()` doesn't support updating primary
> keys (composite or not).
> 
> This means using :class:`~django.forms.ModelForm` with a composite
> primary key may lead to unexpected behavior.
> 
> To avoid issues, make sure :attr:`~django.db.models.Field.editable`
> is set to ``False`` on all primary key fields.

Reviewer 2: ```suggestion
As a composite primary key is a virtual field, a field which doesn't represent
a single database column, this field is excluded from ModelForms::

    class OrderLineItemForm(forms.ModelForm):
        class Meta:
            model = OrderLineItem
            fields = "__all__"

.. code-block:: pycon

    >>> OrderLineItemForm()
    <OrderLineItemForm bound=False, valid=Unknown, fields=(product;order;quantity)>

Setting the primary composite field ``pk`` as a form field raises an unknown
field :exc:`.FieldError`.
```
I feel like forms work fine? I'm not sure the issue trying to be highlighted here. Can you confirm the behavior with tests?

Author: The way it works is `ModelForm` always excludes the primary key.

So yeah, they work. There's a test: `CompositePKTests.test_model_forms`.

The issue I tried to highlight is that the `ModelForm` doesn't exclude the fields of the composite primary key automatically. So those fields are editable on the form, which may lead to unexpected results.

Reviewer 2: I see. I found a note on this within [Field.primary_key](https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.Field.primary_key) (see the "The primary key field is read-only..")
So I have made this an admonition to reference that :+1:

=====================================
How to define a composite primary key
=====================================

.. versionadded:: 5.2

In Django, each model has a primary key. By default, this primary key consists
of a single field.

In most cases, a single primary key should suffice. In database design,
however, defining a primary key consisting of multiple fields is sometimes
necessary.

To define a composite primary key, set the model's ``pk`` to an instance of
``CompositePrimaryKey``:

.. code-block:: pycon

    class Tenant(models.Model):
        pass

    class User(models.Model):
        pk = models.CompositePrimaryKey("tenant_id", "id")
        tenant = models.ForeignKey(Tenant, on_delete=models.CASCADE)
        id = models.IntegerField()

This will instruct Django to create a composite primary key

> (``PRIMARY KEY (tenant_id, id)``) when creating the table.

Author: Okay, we can change this. For me the main reason to use composite pks was for the sharding / partitioning of multi-tenant apps, that's why I was sticking to these examples 😄

It's possible to define a composite primary key consisting of a single field.
This is _almost_ the same as setting ``primary_key=True`` on the field.

.. code-block:: pycon

    class Foo(models.Model):
        pk = models.CompositePrimaryKey("id")

>         id = models.IntegerField()

Author: Or we could prevent them from doing this? I don't really have any arguments for or against doing this.

The only difference is `pk` will be a tuple, and the primary key will be created as a table-level constraint instead of a column-level constraint when creating the table.

>         id = models.IntegerField()

Reviewer 2: I see 🤔 maybe let's not support and then if people give reasons why they want it, we can add support. It's probably more likely to be a mistake and then they can't migrate after

>         id = models.IntegerField()

Author: I added an error for this.

Changes to *pk*
===============

In Django, composite primary keys are represented with ``tuple``\s:

.. code-block:: pycon

    >>> user = User.objects.create(tenant_id=1, id=2)
    >>> user.pk

>     (1, 2)

Author: Sure

        tenant = models.ForeignKey(Tenant, on_delete=models.CASCADE)
        user_id = models.IntegerField()
        user = models.ForeignObject(
            User,
            on_delete=models.CASCADE,
            from_fields=("tenant_id", "user_id"),
            to_fields=("tenant_id", "id"),
        )

``ForeignObject`` is much like ``ForeignKey``,
except that it doesn't create any columns (e.g. ``user_id``), foreign key
constraints or indexes in the database.

.. warning::

    ``ForeignObject`` is an internal API. The work to provide a public API for
    composite foreign keys is already under way. Use it at your own risk.

Composite primary keys and generic relations
============================================

The composite primary key feature is still in its infancy, it doesn't
support generic relations yet.

The work to add composite primary key support to generic relations
is already under way.

Composite primary keys and database functions
=============================================

The database functions may not work with composite primary keys.

This is because most database functions accept a single field only.

.. code-block:: sql

    COUNT("id")  -- ✓ OK

>     COUNT("tenant_id", "id")  -- 𐄂 ERROR

Author: I removed it ✅

=====================================
How to define a composite primary key
=====================================

In Django, each model defines a primary key.
By default, this primary key consists of a single field.

In most cases, a single-field primary key should suffice.
In database design, however, defining a primary key that consists of multiple
fields is sometimes necessary.

To define a composite primary key, set the ``Meta.primary_key`` option:

.. code-block:: pycon

    class Tenant(models.Model):
        pass

    class User(models.Model):
        tenant = models.ForeignKey(Tenant, on_delete=models.CASCADE)
        id = models.IntegerField()

        class Meta:
            primary_key = ("tenant_id", "id")

This will instruct Django to create a composite ``PRIMARY KEY`` constraint
instead of a single-field primary key.

.. note::

    If ``Meta.primary_key`` is set, none of the model's field may set
    ``primary_key=True``.

> 

Author: Yes, it's supported, I think there's no harm in allowing it, although there is no real advantage to it either. WDYT? We can mention it of course, let me add another sentence here.

> 

Reviewer 5: I don't have a great preference. But I think it's worth stating, because folks will ask (I'd wager)

> 

Author: Sure, I added it!

At the moment, relationship fields (``ForeignKey``, ``OneToOneField``,
``ManyToManyField``) don't support composite primary keys.

So, given the ``Tenant`` and ``User`` models, we cannot do:

.. code-block:: pycon

    class Comment(models.Model):
        pk = models.CompositePrimaryKey("tenant_id", "id")
        tenant = models.ForeignKey(Tenant, on_delete=models.CASCADE)

>         id = models.IntegerField()

Reviewer: I think giving `Comment` a composite primary key is a distraction from the point about `ForeignKey`.

Author: Yeah, fair point I adjusted it
